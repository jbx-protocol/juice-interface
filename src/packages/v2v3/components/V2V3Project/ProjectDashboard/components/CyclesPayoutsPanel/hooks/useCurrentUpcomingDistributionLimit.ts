import { useProjectMetadataContext } from 'contexts/ProjectMetadataContext'
import { useProjectContext } from 'packages/v2v3/components/V2V3Project/ProjectDashboard/hooks/useProjectContext'
import useProjectDistributionLimit from 'packages/v2v3/hooks/contractReader/useProjectDistributionLimit'
import { useProjectUpcomingFundingCycle } from 'packages/v2v3/hooks/contractReader/useProjectUpcomingFundingCycle'

export const useCurrentUpcomingDistributionLimit = (
  type: 'current' | 'upcoming',
) => {
  const { projectId } = useProjectMetadataContext()
  const {
    primaryETHTerminal,
    distributionLimit: currentDistributionLimit,
    distributionLimitCurrency: currentDistributionLimitCurrency,
  } = useProjectContext()

  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const {
    data: upcomingDistributionLimitData,
    loading: upcomingDistributionLimitLoading,
  } = useProjectDistributionLimit({
    projectId,
    configuration: upcomingFundingCycle?.configuration.toString(),
    terminal: primaryETHTerminal,
  })

  const [upcomingDistributionLimit, upcomingDistributionLimitCurrency] =
    upcomingDistributionLimitData ?? []

  if (type === 'current') {
    return {
      distributionLimit: currentDistributionLimit,
      distributionLimitCurrency: currentDistributionLimitCurrency,
      loading: false,
    }
  }
  return {
    distributionLimit: upcomingDistributionLimit,
    distributionLimitCurrency: upcomingDistributionLimitCurrency,
    loading: upcomingFundingCycleLoading || upcomingDistributionLimitLoading,
  }
}
