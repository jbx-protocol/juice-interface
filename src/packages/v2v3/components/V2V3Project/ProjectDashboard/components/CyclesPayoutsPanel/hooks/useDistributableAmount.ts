import { BigNumber } from 'ethers'
import { V2V3ProjectContext } from 'packages/v2v3/contexts/Project/V2V3ProjectContext'
import { V2V3CurrencyOption } from 'packages/v2v3/models/currencyOption'
import { V2V3_CURRENCY_ETH } from 'packages/v2v3/utils/currency'
import { useContext } from 'react'

export const useDistributableAmount = () => {
  const {
    distributionLimit,
    usedDistributionLimit,
    balanceInDistributionLimitCurrency,
    distributionLimitCurrency,
  } = useContext(V2V3ProjectContext)

  const currency = (distributionLimitCurrency?.toNumber() ||
    V2V3_CURRENCY_ETH) as V2V3CurrencyOption

  const effectiveDistributionLimit = distributionLimit ?? BigNumber.from(0)
  const distributedAmount = usedDistributionLimit ?? BigNumber.from(0)
  const balInDistributionLimit =
    balanceInDistributionLimitCurrency ?? BigNumber.from(0)

  const distributable = effectiveDistributionLimit.eq(0)
    ? effectiveDistributionLimit
    : effectiveDistributionLimit.sub(distributedAmount)

  const distributableAmount = balInDistributionLimit.gt(distributable)
    ? distributable
    : balInDistributionLimit

  return {
    distributableAmount,
    currency,
  }
}
