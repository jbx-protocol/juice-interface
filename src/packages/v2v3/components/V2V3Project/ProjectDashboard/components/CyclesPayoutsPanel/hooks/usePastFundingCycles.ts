import {
  FundingCycle_OrderBy,
  OrderDirection,
  useFundingCyclesQuery,
} from 'generated/graphql'
import { client } from 'lib/apollo/client'

const DEFAULT_PAGE_SIZE = 5

export function usePastFundingCycles({
  projectId,
  currentFcNumber,
  pageSize,
}: {
  projectId: number | undefined
  currentFcNumber: number | undefined
  pageSize?: number
}) {
  return useFundingCyclesQuery({
    client,
    variables: {
      where: {
        projectId,
        number_lt: currentFcNumber,
      },
      orderBy: FundingCycle_OrderBy.number,
      orderDirection: OrderDirection.desc,
      first: pageSize ?? DEFAULT_PAGE_SIZE,
    },
  })
}
