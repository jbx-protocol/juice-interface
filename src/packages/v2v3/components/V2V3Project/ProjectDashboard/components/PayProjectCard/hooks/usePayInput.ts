import {
  V2V3_CURRENCY_ETH,
  V2V3_CURRENCY_USD,
} from 'packages/v2v3/utils/currency'
import { useState } from 'react'
import { PayInputValue } from '../components/PayInput'

export const usePayInput = (
  v?: PayInputValue,
  oc?: (value: PayInputValue) => void,
) => {
  const [_value, _setValue] = useState<PayInputValue>({
    amount: '',
    currency: V2V3_CURRENCY_ETH,
  })
  const value = v ?? _value
  const onChange = oc ?? _setValue

  const onInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const v = e.target.value
    onChange({
      ...value,
      amount: v,
    })
  }
  const onCurrencyChange = () => {
    onChange({
      ...value,
      currency:
        value.currency === V2V3_CURRENCY_ETH
          ? V2V3_CURRENCY_USD
          : V2V3_CURRENCY_ETH,
    })
  }

  return {
    value,
    onInputChange,
    onCurrencyChange,
  }
}
