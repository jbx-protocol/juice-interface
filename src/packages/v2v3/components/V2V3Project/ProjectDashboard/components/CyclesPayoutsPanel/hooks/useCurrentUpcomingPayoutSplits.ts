import { ETH_PAYOUT_SPLIT_GROUP } from 'constants/splits'
import { useProjectMetadataContext } from 'contexts/ProjectMetadataContext'
import { useProjectContext } from 'packages/v2v3/components/V2V3Project/ProjectDashboard/hooks/useProjectContext'
import useProjectSplits from 'packages/v2v3/hooks/contractReader/useProjectSplits'
import { useProjectUpcomingFundingCycle } from 'packages/v2v3/hooks/contractReader/useProjectUpcomingFundingCycle'

export const useCurrentUpcomingPayoutSplits = (
  type: 'current' | 'upcoming',
) => {
  const { projectId } = useProjectMetadataContext()
  const { payoutSplits: currentPayoutSplits } = useProjectContext()

  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const { data: upcomingPayoutSplits, loading: upcomingProjectSplitsLoading } =
    useProjectSplits({
      projectId,
      splitGroup: ETH_PAYOUT_SPLIT_GROUP,
      domain: upcomingFundingCycle?.configuration?.toString(),
    })

  if (type === 'current') {
    return { splits: currentPayoutSplits, loading: false }
  }
  return {
    splits: upcomingPayoutSplits,
    loading: upcomingFundingCycleLoading || upcomingProjectSplitsLoading,
  }
}
