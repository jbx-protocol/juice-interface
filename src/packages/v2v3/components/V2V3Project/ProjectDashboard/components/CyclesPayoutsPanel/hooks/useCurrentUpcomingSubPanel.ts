import { t } from '@lingui/macro'
import { useProjectMetadataContext } from 'contexts/ProjectMetadataContext'
import { useFundingCycleCountdown } from 'packages/v2v3/components/V2V3Project/ProjectDashboard/hooks/useFundingCycleCountdown'
import { useProjectContext } from 'packages/v2v3/components/V2V3Project/ProjectDashboard/hooks/useProjectContext'
import { timeSecondsToDateString } from 'packages/v2v3/components/V2V3Project/ProjectDashboard/utils/timeSecondsToDateString'
import { useProjectUpcomingFundingCycle } from 'packages/v2v3/hooks/contractReader/useProjectUpcomingFundingCycle'
import { BallotState } from 'packages/v2v3/models/fundingCycle'
import { useMemo } from 'react'

export const useCurrentUpcomingSubPanel = (type: 'current' | 'upcoming') => {
  const { projectId } = useProjectMetadataContext()
  const {
    fundingCycle,
    loading: { fundingCycleLoading },
  } = useProjectContext()
  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({
    projectId,
    /**
     * if the current cycle is unlocked, force the use of latestConfiguredFundingCycleOf.
     */
    useLatestConfigured: fundingCycle?.duration?.isZero() ?? false,
  })
  const [upcomingFundingCycle, , ballotState] = upcomingFundingCycleData ?? []
  const { timeRemainingText } = useFundingCycleCountdown()

  const cycleNumber = useMemo(() => {
    if (type === 'current') {
      return fundingCycle?.number.toNumber()
    }
    return fundingCycle?.number ? fundingCycle.number.toNumber() + 1 : undefined
  }, [fundingCycle?.number, type])

  const cycleUnlocked = useMemo(() => {
    if (type === 'current') {
      return fundingCycle?.duration?.isZero() ?? true
    }
    return upcomingFundingCycle?.duration?.isZero() ?? true
  }, [fundingCycle?.duration, type, upcomingFundingCycle?.duration])

  const upcomingCycleLength = useMemo(() => {
    if (!upcomingFundingCycle) return
    if (cycleUnlocked) return '-'
    return timeSecondsToDateString(
      upcomingFundingCycle.duration.toNumber(),
      'short',
    )
  }, [cycleUnlocked, upcomingFundingCycle])

  /** Determines if the CURRENT cycle is unlocked.
   * This is used to check if the upcoming cycle can start at any time. */
  const currentCycleUnlocked = fundingCycle?.duration?.isZero() ?? true

  const status = cycleUnlocked ? t`Unlocked` : t`Locked`
  const remainingTime = cycleUnlocked ? '-' : timeRemainingText

  // Short circuit current for faster loading
  if (type === 'current') {
    if (fundingCycleLoading) return { loading: true, type }
    return {
      loading: false,
      type,
      cycleNumber,
      status,
      remainingTime,
    }
  }

  if (fundingCycleLoading || upcomingFundingCycleLoading)
    return {
      loading: true,
      type,
    }

  return {
    loading: false,
    type,
    cycleNumber,
    status,
    cycleLength: upcomingCycleLength,
    cycleUnlocked,
    currentCycleUnlocked,
    hasPendingConfiguration:
      /**
       * If a cycle is unlocked, it may have a pending change.
       * The only way it would, is if the ballot state of the latestConfiguredFC is `approved`.
       */
      cycleUnlocked &&
      typeof ballotState !== 'undefined' &&
      ballotState !== null &&
      ballotState === BallotState.approved,
  }
}
