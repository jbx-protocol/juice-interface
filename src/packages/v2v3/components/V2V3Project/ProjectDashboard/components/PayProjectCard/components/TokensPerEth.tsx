import { Trans } from '@lingui/macro'
import { V2V3CurrencyOption } from 'packages/v2v3/models/currencyOption'
import { V2V3_CURRENCY_ETH } from 'packages/v2v3/utils/currency'
import { useTokensPerEth } from '../../../hooks/useTokensPerEth'

export const TokensPerEth = ({
  currencyAmount,
}: {
  currencyAmount:
    | {
        amount: number // NOTE this is actually a `string | undefined` from what I can tell. Too scared to change it atm.
        currency: V2V3CurrencyOption
      }
    | undefined
}) => {
  const { currencyText, receivedTickets, receivedTokenSymbolText } =
    useTokensPerEth({
      amount: parseFloat(currencyAmount?.amount?.toString() || '1'),
      currency: currencyAmount?.currency || V2V3_CURRENCY_ETH,
    })

  const suffix =
    !currencyAmount || !currencyAmount.amount ? (
      <Trans>per {currencyText} paid</Trans>
    ) : (
      ''
    )
  return (
    <Trans>
      Receive {receivedTickets} {receivedTokenSymbolText} {suffix}
    </Trans>
  )
}
