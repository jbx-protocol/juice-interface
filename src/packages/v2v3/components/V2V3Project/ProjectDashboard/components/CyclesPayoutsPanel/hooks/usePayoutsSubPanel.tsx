import { AmountInCurrency } from 'components/currency/AmountInCurrency'
import { Split } from 'models/splits'
import { useProjectContext } from 'packages/v2v3/components/V2V3Project/ProjectDashboard/hooks/useProjectContext'
import { V2V3CurrencyOption } from 'packages/v2v3/models/currencyOption'
import { V2V3CurrencyName } from 'packages/v2v3/utils/currency'
import { isJuiceboxProjectSplit } from 'packages/v2v3/utils/distributions'
import { isInfiniteDistributionLimit } from 'packages/v2v3/utils/fundingCycle'
import {
  SPLITS_TOTAL_PERCENT,
  feeForAmount,
  formatSplitPercent,
} from 'packages/v2v3/utils/math'
import { useCallback, useMemo } from 'react'
import assert from 'utils/assert'
import { getProjectOwnerRemainderSplit } from 'utils/splits'
import { useCurrentUpcomingDistributionLimit } from './useCurrentUpcomingDistributionLimit'
import { useCurrentUpcomingPayoutSplits } from './useCurrentUpcomingPayoutSplits'
import { useDistributableAmount } from './useDistributableAmount'

const splitHasFee = (split: Split) => {
  return !isJuiceboxProjectSplit(split)
}

const calculateSplitAmountWad = (
  split: Split,
  distributionLimit: bigint | undefined,
  primaryETHTerminalFee: bigint | undefined,
) => {
  const splitValue = distributionLimit
    ? (distributionLimit * BigInt(split.percent)) / SPLITS_TOTAL_PERCENT
    : undefined
  const feeAmount = splitHasFee(split)
    ? feeForAmount(splitValue, primaryETHTerminalFee) ?? BigInt(0)
    : BigInt(0)
  return splitValue ? splitValue - feeAmount : undefined
}

export const usePayoutsSubPanel = (type: 'current' | 'upcoming') => {
  const { splits, loading } = useCurrentUpcomingPayoutSplits(type)
  const { projectOwnerAddress, primaryETHTerminalFee } = useProjectContext()
  const { distributableAmount } = useDistributableAmount()

  const { distributionLimit, distributionLimitCurrency } =
    useCurrentUpcomingDistributionLimit(type)

  const showAmountOnPayout =
    distributionLimit === 0n || isInfiniteDistributionLimit(distributionLimit)
      ? false
      : true

  const transformSplit = useCallback(
    (split: Split) => {
      assert(split.beneficiary, 'Beneficiary must be defined')
      let amount = undefined
      const splitAmountWad = calculateSplitAmountWad(
        split,
        distributionLimit,
        primaryETHTerminalFee,
      )
      if (showAmountOnPayout && splitAmountWad && distributionLimitCurrency) {
        amount = (
          <AmountInCurrency
            amount={splitAmountWad}
            currency={V2V3CurrencyName(
              Number(distributionLimitCurrency) as V2V3CurrencyOption,
            )}
          />
        )
      }
      return {
        projectId: split.projectId
          ? Number(BigInt(split.projectId))
          : undefined,
        address: split.beneficiary!,
        percent: `${formatSplitPercent(BigInt(split.percent))}%`,
        amount,
      }
    },
    [
      distributionLimit,
      distributionLimitCurrency,
      primaryETHTerminalFee,
      showAmountOnPayout,
    ],
  )

  const totalPayoutAmount = useMemo(() => {
    if (!distributionLimit || !distributionLimitCurrency) return
    if (
      isInfiniteDistributionLimit(distributionLimit) ||
      distributionLimit === 0n
    )
      return

    return (
      <AmountInCurrency
        amount={distributionLimit}
        currency={V2V3CurrencyName(
          Number(distributionLimitCurrency) as V2V3CurrencyOption,
        )}
      />
    )
  }, [distributionLimit, distributionLimitCurrency])

  const payouts = useMemo(() => {
    if (loading || !splits) return

    if (
      // We don't need to worry about upcoming as this is informational only
      type === 'current' &&
      splits.length === 0 &&
      distributableAmount === 0n
    ) {
      return []
    }

    const ownerPayout = projectOwnerAddress
      ? getProjectOwnerRemainderSplit(projectOwnerAddress, splits)
      : undefined

    return [
      ...splits,
      ...(ownerPayout && ownerPayout.percent > 0 ? [ownerPayout] : []),
    ]
      .sort((a, b) => Number(b.percent) - Number(a.percent))
      .map(transformSplit)
  }, [
    distributableAmount,
    loading,
    projectOwnerAddress,
    splits,
    transformSplit,
    type,
  ])

  return {
    loading,
    payouts,
    totalPayoutAmount,
    distributionLimit,
  }
}
