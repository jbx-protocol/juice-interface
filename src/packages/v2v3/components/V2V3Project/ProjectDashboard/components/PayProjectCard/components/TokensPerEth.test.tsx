/**
 * @jest-environment jsdom
 */
import { render, screen } from '@testing-library/react'
import { useTokensPerEth } from 'packages/v2v3/components/V2V3Project/ProjectDashboard/hooks/useTokensPerEth'
import { V2V3_CURRENCY_ETH } from 'packages/v2v3/utils/currency'
import { TokensPerEth } from './TokensPerEth'

jest.mock(
  'packages/v2v3/components/V2V3Project/ProjectDashboard/hooks/useTokensPerEth',
  () => ({
    useTokensPerEth: jest.fn(),
  }),
)

describe('TokensPerEth', () => {
  const mockedTokensPerEth = {
    receivedTickets: '100',
    currencyText: 'ETH',
    receivedTokenSymbolText: 'TKN',
  }

  beforeEach(() => {
    ;(useTokensPerEth as jest.Mock).mockReturnValue(mockedTokensPerEth)
  })

  it('renders when `currencyAmount` is 1', () => {
    const { container } = render(
      <TokensPerEth
        currencyAmount={{
          amount: 1,
          currency: V2V3_CURRENCY_ETH,
        }}
      />,
    )
    expect(container).toMatchSnapshot()
    expect(screen.getByText('Receive 100 TKN')).toBeInTheDocument()
  })

  it('calls useTokensPerEth with correct arguments', () => {
    render(
      <TokensPerEth
        currencyAmount={{
          amount: 1,
          currency: V2V3_CURRENCY_ETH,
        }}
      />,
    )
    expect(useTokensPerEth).toHaveBeenCalledWith({
      amount: 1,
      currency: V2V3_CURRENCY_ETH,
    })
  })
})
