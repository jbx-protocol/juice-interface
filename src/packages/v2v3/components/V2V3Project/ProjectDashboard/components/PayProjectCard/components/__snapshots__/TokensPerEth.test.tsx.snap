// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokensPerEth renders when \`currencyAmount\` is 1 1`] = `
<div>
  Receive 100 TKN 
</div>
`;
